src/python/pants/backend/python/util_rules/pex_from_targets.py

```diff
@@ -327,10 +327,37 @@ async def global_requirement_constraints(
     )
 
 
+@dataclass(frozen=True)
+class _PexRequirementsRequest:
+    """Determine the requirement strings used transitively.
+
+    This type is private because callers should likely use `RequirementsPexRequest` or
+    `PexFromTargetsRequest` instead.
+    """
+
+    addresses: Addresses
+
+
+@rule
+async def determine_requirement_strings_in_closure(
+    request: _PexRequirementsRequest, global_requirement_constraints: GlobalRequirementConstraints
+) -> PexRequirements:
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(request.addresses))
+    return PexRequirements.create_from_requirement_fields(
+        (
+            tgt[PythonRequirementsField]
+            for tgt in transitive_targets.closure
+            if tgt.has_field(PythonRequirementsField)
+        ),
+        constraints_strings=(str(constraint) for constraint in global_requirement_constraints),
+    )
+
+
 @frozen_after_init
 @dataclass(unsafe_hash=True)
 class _RepositoryPexRequest:
     addresses: Addresses
+    requirements: PexRequirements
```

Comment:
It would be good for docstring or field naming to clarify that both `addresses` and `requirements` are subsets (afaict) of the entire resolve, only used for staleness checks.

My first read of this (without thinking too much about the review title) was that we were passing in all of the requirements that made up of the resolve (which felt upside down, and I was about to comment to that effect).

Reply:
Agreed the names in general are confusing. I think `Lockfile(req_strings=)` is also confusing. I'm going to punt on this for a holistic rename: https://github.com/pantsbuild/pants/issues/14388

It's a little tricky that interpreter constraints _do_ change how the _RepositoryPex is used, and `addresses` is used to determine those.